lib/rtti.h

```diff
@@ -0,0 +1,297 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef LIB_RTTI_H_
+#define LIB_RTTI_H_
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <type_traits>
+
+/// @file
+/// Custom intrusive lightweight RTTI implementation for semi-open class hierarchies.
+///
+/// P4C IR::Node class herarchy relies a lot on virtual bases and at the same
+/// time does a lot of dynamic type checks and downcasts /
+/// upcasts. Unfortunately, `static_cast` cannot be used to cast to / via
+/// virtual bases and use of `dynamic_cast` involves a non-negligible overhead.
+/// The typical IR::Node class hierarchy looks like this:
+///   class Node : public virtual INode;
+///   class StatOrDecl : public Node;
+///   class IDeclaration : public virtual INode;
+///   class Declaration : public StatOrDecl, public virtual IDeclaration;
+/// and it is pretty common code pattern that does something like this:
+///   IR:: Node *n = new Declaration();
```

Comment:
Inconsistent use of `IR::`

```suggestion
///   Node *n = new Declaration();